README.md
# Filecoin Address

## Usage

Eth address to f4 address
```dart
final address = '0xd4c5fb16488aa48081296299d54b0c648c9333da';
final ethAddress = FilEthAddress.fromEthString(address);
print(ethAddress.toString());
```

f4 address to Eth address
```dart
final address = 'f410f2tc7wfsirksibajjmkm5ksymmsgjgm62hjnomwa';
final ethAddress = FilEthAddress.fromString(address);
print(ethAddress.toEthString());
```

CHANGELOG.md
## 1.1.0

- Support all address types.

## 1.0.1

- Update readme.

## 1.0.0

- Initial version.


test/filecoin_address_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';

import 'package:filecoin_address/filecoin_address.dart';
import 'package:filecoin_address/src/utils.dart';

class AddressData {
  String bytes;
  String string;
  String network;
  int protocol;
  String payload;
  AddressData({
    required this.bytes,
    required this.string,
    required this.network,
    required this.protocol,
    required this.payload,
  });

  Map<String, dynamic> toMap() {
    final result = <String, dynamic>{};

    result.addAll({'bytes': bytes});
    result.addAll({'string': string});
    result.addAll({'network': network});
    result.addAll({'protocol': protocol});
    result.addAll({'payload': payload});

    return result;
  }

  factory AddressData.fromMap(Map<String, dynamic> map) {
    return AddressData(
      bytes: map['bytes'] ?? '',
      string: map['string'] ?? '',
      network: map['network'] ?? '',
      protocol: map['protocol']?.toInt() ?? 0,
      payload: map['payload'] ?? '',
    );
  }

  String toJson() => json.encode(toMap());

  factory AddressData.fromJson(String source) =>
      AddressData.fromMap(json.decode(source));
}

void main() {
  group('Convert evm address', () {
    test('Checksum', () {
      final payload = '040ad4c5fb16488aa48081296299d54b0c648c9333da';
      final result = '3a5ae658';
      final hash = getChecksumForPayload(hexToUint8List(payload));
      expect(listToHex(hash), result);
    });
    test('Eth address to f4', () {
      final address = '0xd4c5fb16488aa48081296299d54b0c648c9333da';
      final ethAddress = FilEthAddress.fromEthString(address);
      expect(ethAddress.toString(),
          'f410f2tc7wfsirksibajjmkm5ksymmsgjgm62hjnomwa');
    });
    test('f4 to Eth address', () {
      final address = 'f410f2tc7wfsirksibajjmkm5ksymmsgjgm62hjnomwa';
      final ethAddress = FilEthAddress.fromString(address);
      expect(ethAddress.toString(), address);
      expect(ethAddress.toEthString(),
          '0xd4c5fb16488aa48081296299d54b0c648c9333da');
    });
    test('Testnet', () {
      final addr = Address.fromString('t08666');
      expect(addr.toString(), 't08666');
      expect(listToHex(addr.toBytes()), '00da43');
      expect(addr.protocol, ProtocolIndicator.ID);
      expect(addr.networkPrefix, 't');
      expect(Address.isAddressId(addr), true);
      if (addr is AddressId) expect(addr.id, '8666');
    });

    test('Mainnet', () {
      final addr = Address.fromString('f08666');
      expect(addr.toString(), 'f08666');
      expect(listToHex(addr.toBytes()), '00da43');
      expect(addr.protocol, ProtocolIndicator.ID);
      expect(addr.networkPrefix, 'f');
      expect(Address.isAddressId(addr), true);
      if (addr is AddressId) expect(addr.id, '8666');
    });
    test('Exceed max id (super big)', () {
      try {
        Address.fromString('t0111111111111111111111111');
      } catch (e) {
        expect(e.toString(), 'Invalid address');
      }
    });

    test('To ethereum address (ID)', () async {
      final addr = Address.fromString('f0101');
      if (addr is AddressId) {
        expect(addr.toEthAddressHex(true),
            '0xff00000000000000000000000000000000000065');
        expect(addr.toEthAddressHex(false),
            'ff00000000000000000000000000000000000065');
      }
    });
  });

  group("Vectors", () {
    group("fromString", () {
      final vectors = jsonDecode(File('test/addresses.json').readAsStringSync());
      final testAdressed = [for (final v in vectors) AddressData.fromMap(v)];
      for (var item in testAdressed) {
        test('formAddress', () {
          final addr = Address.fromString(item.string);
          expect(addr.toString(), item.string);
          expect(listToHex(addr.toBytes()), item.bytes);
          expect(addr.protocol.index, item.protocol);
          expect(addr.networkPrefix, item.network);
          expect(listToHex(addr.payload), item.payload);

          if (addr is AddressId) expect(addr.id, item.string.substring(2));
          if (addr is AddressDelegated) {
            expect(addr.namespace,
                item.string.substring(2, item.string.indexOf(item.network, 1)));
          }
        });
      }
    });
  });
}


lib/src/utils.dart
// ignore_for_file: constant_identifier_names

import 'dart:typed_data';
import 'package:base32/base32.dart';
import 'package:hashlib/hashlib.dart';

Uint8List getChecksumForPayload(Uint8List payload) {
  final blake2b = Blake2b(4);
  final sink = blake2b.createSink();
  sink.add(payload);
  return sink.digest().buffer.asUint8List();
}

bool validateNetworkPrefix(String networkPrefix) {
  return ['f', 't'].contains(networkPrefix);
}

Uint8List hexToUint8List(String hexString) {
  final hexString0 =
      hexString.startsWith('0x') ? hexString.substring(2) : hexString;
  final hash = List<int>.generate(hexString0.length ~/ 2,
      (i) => int.parse(hexString0.substring(i * 2, i * 2 + 2), radix: 16));
  return Uint8List.fromList(hash);
}

String listToHex(List<int> bytes) {
  StringBuffer buffer = StringBuffer();
  for (int part in bytes) {
    if (part & 0xff != part) {
      throw FormatException("Non-byte integer detected");
    }
    buffer.write('${part < 16 ? '0' : ''}${part.toRadixString(16)}');
  }
  return buffer.toString();
}

String base32Encode(Uint8List input) {
  var encoded = base32.encode(input);
  while (encoded.endsWith('=')) {
    encoded = encoded.substring(0, encoded.length - 1);
  }
  return encoded;
}


lib/src/filecoin_address_base.dart
import 'dart:typed_data';

import 'package:base32/base32.dart';
import 'package:leb128/leb128.dart';

import 'utils.dart';
import 'constant.dart';
import 'error.dart';

// ignore: constant_identifier_names
enum ProtocolIndicator { ID, SECP256K1, ACTOR, BLS, DELEGATED }

class Address {
  late final Uint8List _payload;
  late final String _networkPrefix;
  ProtocolIndicator protocol = ProtocolIndicator.SECP256K1;

  String get networkPrefix => _networkPrefix;
  Uint8List get payload => _payload;

  Uint8List toBytes() {
    return Uint8List.fromList([protocol.index] + _payload);
  }

  Uint8List getChecksum() {
    return getChecksumForPayload(toBytes());
  }

  static Address fromString(String address) {
    final type = int.parse(address[1]);
    final protocol = ProtocolIndicator.values[type];
    switch (protocol) {
      case ProtocolIndicator.ID:
        return AddressId.fromAddressString(address);
      case ProtocolIndicator.ACTOR:
        return AddressActor.fromAddressString(address);
      case ProtocolIndicator.SECP256K1:
        return AddressSecp256k1.fromAddressString(address);
      case ProtocolIndicator.BLS:
        return AddressBls.fromAddressString(address);
      case ProtocolIndicator.DELEGATED:
        final addr = AddressDelegated.fromAddressString(address);
        if (Address.isFilEthAddress(addr)) {
          return FilEthAddress(addr.networkPrefix, addr.subAddress);
        }
        return addr;
      default:
        throw FilAddressError("Unknown protocol");
    }
  }

  static bool isAddressId(Address address) =>
      address.protocol == ProtocolIndicator.ID;

  static bool isAddressBls(Address address) =>
      address.protocol == ProtocolIndicator.BLS;

  static bool isAddressActor(Address address) =>
      address.protocol == ProtocolIndicator.ACTOR;

  static bool isAddressSecp256k1(Address address) =>
      address.protocol == ProtocolIndicator.SECP256K1;

  static bool isAddressDelegate(Address address) =>
      address.protocol == ProtocolIndicator.DELEGATED;

  static bool isFilEthAddress(Address address) {
    return address is AddressDelegated && address.namespace == '10';
  }
}

class AddressId extends Address {
  late final String _id;
  AddressId(String networkPrefix, Uint8List payload) {
    if (payload.length > ID_PAYLOAD_MAX_LEN) {
      throw FilAddressError("Invalid ID address payload length");
    }
    _payload = payload;
    _networkPrefix = networkPrefix;
    _id = toString().substring(2);
    protocol = ProtocolIndicator.ID;
  }

  String get id => _id;

  static AddressId fromAddressString(String address) {
    final networkPrefix = address[0];
    final protocolIndicator = address[1];

    if (!validateNetworkPrefix(networkPrefix)) {
      throw FilAddressError("Invalid network prefix");
    }

    if (int.parse(protocolIndicator) != ProtocolIndicator.ID.index) {
      throw FilAddressError("Invalid protocol indicator");
    }
    final addressInt = int.tryParse(address.substring(2));
    if (addressInt == null) throw FilAddressError("Invalid address");
    final payload = Leb128.encodeUnsigned(addressInt);
    return AddressId(networkPrefix, payload);
  }

  String toEthAddressHex([bool hexPrefix = false]) {
    final payloadString = listToHex(_payload);
    return '${hexPrefix ? '0xff' : 'ff'}${payloadString.padLeft(ETH_ADDRESS_LEN * 2, '0')}';
  }

  @override
  String toString() =>
      networkPrefix +
      protocol.index.toString() +
      Leb128.decodeUnsigned(payload).toString();
}

class AddressActor extends Address {
  AddressActor(String networkPrefix, Uint8List payload) {
    _networkPrefix = networkPrefix;
    if (payload.length > ACTOR_PAYLOAD_LEN) {
      throw FilAddressError("Invalid Actor address payload length");
    }
    _payload = payload;
    protocol = ProtocolIndicator.ACTOR;
  }

  static AddressActor fromAddressString(String address) {
    final networkPrefix = address[0];
    final protocolIndicator = address[1];

    if (!validateNetworkPrefix(networkPrefix)) {
      throw FilAddressError("Invalid network prefix");
    }
    if (int.parse(protocolIndicator) != ProtocolIndicator.ACTOR.index) {
      throw FilAddressError("Invalid protocol indicator");
    }

    final decodedData = base32.decode(address.substring(2).toUpperCase());
    final payload = decodedData.sublist(0, decodedData.length - 4);
    final checksum = decodedData.sublist(decodedData.length - 4);

    final newAddress = AddressActor(networkPrefix, payload);
    if (listToHex(newAddress.getChecksum()) != listToHex(checksum)) {
      throw FilAddressError('Invalid checksum address');
    }
    return newAddress;
  }

  @override
  String toString() {
    final checksum = getChecksum();
    return networkPrefix +
        protocol.toString() +
        base32Encode(Uint8List.fromList(payload + checksum)).toLowerCase();
  }
}

class AddressBls extends Address {
  AddressBls(String networkPrefix, Uint8List payload) {
    _networkPrefix = networkPrefix;
    _payload = payload;
    if (payload.length > BLS_PAYLOAD_LEN) {
      throw FilAddressError("Invalid bls address payload length");
    }
    protocol = ProtocolIndicator.BLS;
  }

  static AddressBls fromAddressString(String address) {
    final networkPrefix = address[0];
    final protocolIndicator = address[1];

    if (!validateNetworkPrefix(networkPrefix)) {
      throw FilAddressError("Invalid network prefix");
    }
    if (int.parse(protocolIndicator) != ProtocolIndicator.BLS.index) {
      throw FilAddressError('Invalid proroocol indicator');
    }

    final decodedData = base32.decode(address.substring(2).toUpperCase());
    final payload = decodedData.sublist(0, decodedData.length - 4);
    final checksum = decodedData.sublist(decodedData.length - 4);

    final newAddress = AddressBls(networkPrefix, payload);
    if (listToHex(newAddress.getChecksum()) != listToHex(checksum)) {
      throw FilAddressError('Invalid checksum address');
    }
    return newAddress;
  }

  @override
  String toString() {
    final checksum = getChecksum();
    return networkPrefix +
        protocol.index.toString() +
        base32Encode(Uint8List.fromList(payload + checksum)).toLowerCase();
  }
}

class AddressSecp256k1 extends Address {
  AddressSecp256k1(String networkPrefix, Uint8List payload) {
    _networkPrefix = networkPrefix;
    _payload = payload;
    protocol = ProtocolIndicator.SECP256K1;
  }

  static AddressSecp256k1 fromAddressString(String address) {
    final networkPrefix = address[0];
    final protocolIndicator = address[1];

    if (!validateNetworkPrefix(networkPrefix)) {
      throw FilAddressError("Invalid network prefix");
    }
    if (int.tryParse(protocolIndicator) != 1) {
      throw FilAddressError('Invalid proroocol indicator');
    }

    final decodedData = base32.decode(address.substring(2).toUpperCase());
    final payload = decodedData.sublist(0, decodedData.length - 4);
    final checksum = decodedData.sublist(decodedData.length - 4);

    final newAddress = AddressSecp256k1(networkPrefix, payload);
    if (listToHex(newAddress.getChecksum()) != listToHex(checksum)) {
      throw FilAddressError('Invalid checksum address');
    }
    return newAddress;
  }

  @override
  String toString() {
    final checksum = getChecksum();
    return networkPrefix +
        protocol.index.toString() +
        base32Encode(Uint8List.fromList(payload + checksum)).toLowerCase();
  }
}

class AddressDelegated extends Address {
  late String _namespace;
  late Uint8List _subAddress;
  AddressDelegated(
      String networkPrefix, String namespace, Uint8List subAddress) {
    _networkPrefix = networkPrefix;
    if (BigInt.parse(namespace) > ID_PAYLOAD_MAX_NUM) {
      throw FilAddressError('Invalid namespace');
    }
    if (subAddress.isEmpty || subAddress.length > SUB_ADDRESS_MAX_LEN) {
      throw FilAddressError('Invalid sub address length');
    }
    _namespace = namespace;
    _subAddress = subAddress;
    protocol = ProtocolIndicator.DELEGATED;
    _payload = toBytes().sublist(1);
  }

  String get namespace => _namespace;
  Uint8List get subAddress => _subAddress;

  static AddressDelegated fromAddressString(String address) {
    final networkPrefix = address[0];
    final protocolIndicator = address[1];

    if (!validateNetworkPrefix(networkPrefix)) {
      throw FilAddressError("Invalid network prefix");
    }
    if (int.tryParse(protocolIndicator) != 4) {
      throw FilAddressError('Invalid proroocol indicator');
    }
    final namespace = address.substring(2, address.indexOf('f', 2));
    final decodedData = base32
        .decode(address.substring(address.indexOf('f', 2) + 1).toUpperCase());
    final subAddress = decodedData.sublist(0, decodedData.length - 4);
    final checksum = decodedData.sublist(decodedData.length - 4);

    final newAddress = AddressDelegated(networkPrefix, namespace, subAddress);
    if (listToHex(newAddress.getChecksum()) != listToHex(checksum)) {
      throw FilAddressError('Invalid checksum address');
    }
    return newAddress;
  }

  @override
  Uint8List toBytes() {
    final namespaceBytes = Leb128.encodeUnsigned(int.parse(namespace));
    final protocolBytes = Leb128.encodeUnsigned(protocol.index);

    return Uint8List.fromList(protocolBytes + namespaceBytes + _subAddress);
  }

  @override
  String toString() {
    final checksum = getChecksum();
    return ('$networkPrefix${protocol.index}${_namespace}f${base32Encode(Uint8List.fromList(_subAddress + checksum)).toLowerCase()}');
  }
}

class FilEthAddress extends Address {
  ProtocolIndicator protocolIndicator = ProtocolIndicator.DELEGATED;
  late Uint8List _subAddress;
  final nameSpace = 10;

  FilEthAddress(String networkPrefix, Uint8List ethAddress) {
    _networkPrefix = networkPrefix;
    _subAddress = ethAddress;
    _payload = toBytes().sublist(1);
    protocol = ProtocolIndicator.DELEGATED;
  }
  @override
  Uint8List toBytes() {
    final namespaceBytes = Leb128.encodeUnsigned(nameSpace);
    final protocolBytes = Leb128.encodeUnsigned(4);

    return Uint8List.fromList(protocolBytes + namespaceBytes + _subAddress);
  }

  /// Convert to f4 address.
  @override
  String toString() {
    final checksum = getChecksumForPayload(toBytes());
    return '$networkPrefix${protocol.index}${nameSpace}f${base32Encode(Uint8List.fromList(_subAddress + checksum)).toLowerCase()}';
  }

  /// FilEthAddress from f4 address.
  static FilEthAddress fromString(String address) {
    final networkPrefix = address[0];
    final protocolIndicator = address[1];
    if (int.parse(protocolIndicator) != 4) throw Error();
    final dataEncoded = address.substring(address.indexOf('f', 2) + 1);
    final dataDecoded = base32.decode(dataEncoded.toUpperCase());
    final subAddress = dataDecoded.sublist(0, dataDecoded.length - 4);
    final newAddress = FilEthAddress(networkPrefix, subAddress);
    return newAddress;
  }

  /// FilEthAddress from evm address, 0x prefixed.
  static FilEthAddress fromEthString(String address, [networkPrefix = 'f']) {
    final ethAddress = hexToUint8List(address);
    return FilEthAddress(networkPrefix, ethAddress);
  }

  /// Convert to eth string.
  String toEthString([hexPrefix = true]) {
    return '${hexPrefix ? '0x' : ''}${_subAddress.map((int a) => a.toRadixString(16).padLeft(2, '0')).join('')}';
  }
}
